evap/staff/views.py

```diff
@@ -1002,41 +1015,46 @@ def course_copy(request, course_id):
+    model = Course
+    pk_url_kwarg = "course_id"
+    form_class = CourseForm
+    template_name = "staff_course_form.html"
+    success_message = gettext_lazy("Successfully updated course.")
 
-    course_form = CourseForm(request.POST or None, instance=course)
-    editable = course.can_be_edited_by_manager
-
-    if request.method == "POST" and not editable:
-        raise SuspiciousOperation("Modifying this course is not allowed.")
-
-    operation = request.POST.get("operation")
-
-    if course_form.is_valid():
-        if operation not in ("save", "save_create_evaluation", "save_create_single_result"):
-            raise SuspiciousOperation("Invalid POST operation")
+    object: Course
 
-        if course_form.has_changed():
-            course = course_form.save()
-            update_template_cache_of_published_evaluations_in_course(course)
+    def get_object(self, *args, **kwargs):
+        course = super().get_object(*args, **kwargs)
+        if self.request.method == "POST" and not course.can_be_edited_by_manager:
+            raise SuspiciousOperation("Modifying this course is not allowed.")
+        return course
 
-        messages.success(request, _("Successfully updated course."))
-        if operation == "save_create_evaluation":
-            return redirect("staff:evaluation_create_for_course", course.id)
-        if operation == "save_create_single_result":
-            return redirect("staff:single_result_create_for_course", course.id)
+    def get_context_data(self, **kwargs):
+        context_data = super().get_context_data(**kwargs) | {
+            "semester": self.object.semester,
+            "editable": self.object.can_be_edited_by_manager,
+            "disable_breadcrumb_course": True,
+        }
+        context_data["course_form"] = context_data.pop("form")
+        return context_data
 
-        return redirect("staff:semester_view", course.semester.id)
+    def form_valid(self, form):
+        response = super().form_valid(form)
+        if form.has_changed():
+            update_template_cache_of_published_evaluations_in_course(self.object)
+        return response
 
-    template_data = {
-        "course": course,
-        "semester": course.semester,
-        "course_form": course_form,
-        "editable": editable,
-        "disable_breadcrumb_course": True,
-    }
-    return render(request, "staff_course_form.html", template_data)
+    def get_success_url(self):
+        match self.request.POST.get("operation"):
+            case "save":
+                return reverse("staff:semester_view", args=[self.object.semester.id])
+            case "save_create_evaluation":
+                return reverse("staff:evaluation_create_for_course", args=[self.object.id])
+            case "save_create_single_result":
+                return reverse("staff:single_result_create_for_course", args=[self.object.id])
+            case _:
+                raise SuspiciousOperation("Invalid POST operation")
```

Comment:
This would still have changed the course, and then return the 400 after that, right? Shouldn't be a bit deal, but might be a bit weird.

Reply:
Yes, we could also check it in `form_valid` before calling `super().form_valid` where it would be saved